chain-signatures/node/src/indexer_eth.rs

```diff
@@ -418,23 +512,185 @@ pub async fn run(
+            .with_label_values(&[Chain::Ethereum.as_str(), near_account_id_clone.as_str()])
             .set(block_number as i64);
     }
     Ok(())
 }
 
+async fn retry_failed_blocks(
+    mut blocks_failed_rx: mpsc::Receiver<BlockNumberAndHash>,
+    blocks_failed_tx: mpsc::Sender<BlockNumberAndHash>,
+    client: &Arc<EthereumClient<FileDB>>,
+    eth_contract_addr: Address,
+    node_near_account_id: &AccountId,
+    requests_indexed: mpsc::Sender<BlockAndRequests>,
+) {
+    loop {
+        let Some((block_number, block_hash)) = blocks_failed_rx.recv().await else {
+            tracing::warn!("Failed to receive block and requests from requests_indexed");
+            break;
+        };
+        if let Err(err) = process_block(
+            block_number,
+            block_hash,
+            client,
+            eth_contract_addr,
+            node_near_account_id,
+            requests_indexed.clone(),
+        )
+        .await
+        {
+            tracing::warn!("Retry failed for block {block_number}: {:?}", err);
+            add_failed_block(blocks_failed_tx.clone(), block_number, block_hash).await;
+        } else {
+            tracing::info!("Successfully retried block: {}", block_number);
+        }
+    }
+}
+
+async fn add_failed_block(
+    blocks_failed: mpsc::Sender<BlockNumberAndHash>,
+    block_number: u64,
+    block_hash: alloy::primitives::B256,
+) {
+    blocks_failed
+        .send((block_number, block_hash))
+        .await
+        .unwrap_or_else(|err| {
+            tracing::warn!("Failed to send failed block: {:?}", err);
+        });
+}
+
+/// Polls for the latest finalized block and update the latest finalized epoch.
+/// This will keep polling and only return when there's an error or the final block number and finalized epock is updated.
+/// How it works:
+/// Once finalized block gets updated, we fetch the blocks in the epoch ending with the finalized block.
+/// To ensure the blocks we fetched are proven, we fetch the latest finalized block from Helios.
+/// Then using an unstrusted RPC client, we go backwards to fetch the blocks in the epoch and check that the hash of the block's header == its next block's parent_hash.
+/// This is proven because the finalized block is already proven by helios, and by checking current block hash == next block's parent hash, we prove that each block is indeed included in the chain.
+async fn refresh_finalized_epoch(
+    helios_client: &Arc<EthereumClient<FileDB>>,
+    untrusted_rpc_client: &RootProvider<Http<AlloyClient>>,
+    last_final_block_number: Option<u64>,
+    refresh_finalized_interval: u64,
+) -> anyhow::Result<(Option<u64>, BlockNumberToHashMap)> {
+    let mut interval = tokio::time::interval(Duration::from_millis(refresh_finalized_interval));
+    let mut finalized_epoch = BlockNumberToHashMap::new();
+    let mut final_block_number: Option<u64> = last_final_block_number;
+    loop {
+        interval.tick().await;
```

Comment:
I think we can get rid of the loop. Instead, try once and return afterwards, even if there was no update.

This also moves the interval up one layer, so it's less hidden away.